libraries/AP_GPS/AP_GPS_FEMTO.h

```diff
@@ -0,0 +1,173 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//  Femtomes GPS driver for ArduPilot.
+//  Code by Rui Zheng <ruizheng@femtomes.com>
+
+#pragma once
+
+#include "AP_GPS.h"
+#include "GPS_Backend.h"
+
+class AP_GPS_FEMTO : public AP_GPS_Backend
+{
+public:
+    AP_GPS_FEMTO(AP_GPS &_gps, AP_GPS::GPS_State &_state, AP_HAL::UARTDriver *_port);
+
+    AP_GPS::GPS_Status highest_supported_status(void) override { return AP_GPS::GPS_OK_FIX_3D_RTK_FIXED; }
+
+    // Methods
+    bool read() override;
+
+    void inject_data(const uint8_t *data, uint16_t len) override;
+
+    const char *name() const override { return "FEMTO"; }
+
+private:
+    bool parse(uint8_t temp);
+    bool process_message();
+    uint32_t CRC32Value(uint32_t icrc);
+    uint32_t CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc);
+
+    static const uint8_t FEMTO_PREAMBLE1 = 0xaa;
+    static const uint8_t FEMTO_PREAMBLE2 = 0x44;
+    static const uint8_t FEMTO_PREAMBLE3 = 0x12;
+    
+    static const char* const _initialisation_blob[3];
+   
+    uint32_t crc_error_counter = 0;
```

Comment:
you don't need to init to 0 on ArduPilot, that is done by default, so it is better to remove the zero initialisation to save some flash space